src/Source.zig

```diff
@@ -36,3 +37,25 @@ pub fn lineColString(source: Source, byte_offset: u32) LCS {
     }
     return .{ .line = line, .col = col, .str = source.buf[start..i] };
 }
+
+/// Returns the first offset, if any, in buf where an invalid utf8 sequence
+/// is found. Code adapted from std.unicode.utf8ValidateSlice
+fn offsetOfInvalidUtf8(buf: []const u8) ?u32 {
+    std.debug.assert(buf.len <= std.math.maxInt(u32));
+    var i: u32 = 0;
+    while (i < buf.len) {
+        if (std.unicode.utf8ByteSequenceLength(buf[i])) |cp_len| {
+            if (i + cp_len > buf.len) return i;
+            if (std.meta.isError(std.unicode.utf8Decode(buf[i .. i + cp_len]))) return i;
+            i += cp_len;
+        } else |_| return i;
+    }
+    return null;
+}
+
+pub fn locationOfInvalidUtf8(source: Source) ?Location {
+    if (offsetOfInvalidUtf8(source.buf)) |offset| {
```

Comment:
Not sure how big of an impact it will have but caching the result of this check would be a really simple enhancement to make.

Reply:
are you thinking something like `is_utf8: ?bool` as a member of `Source`?

Reply:
nm, I see we could just store the location itself. Just need to distinguish "haven't checked yet" and "does not have invalid utf8"